File: Vuforia Spatial Toolbox/VideoRecordingManager.m
//
//  VideoRecordingManager.m
//  Vuforia Spatial Toolbox
//
//  Created by Benjamin Reynolds on 5/8/19.
//  Copyright © 2019 PTC. All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//

#import "VideoRecordingManager.h"
#import "FileManager.h"

#define temporaryDirectory NSTemporaryDirectory()

@implementation VideoRecordingManager
{
    BOOL isRecording;
    CFTimeInterval recordingStartTime;
    CFTimeInterval firstFrameOffset;
    NSTimer* updateTimer;
}

// note that the videoRecordingDelegate is not set in the constructor, it MUST be manually set before startRecording is used
+ (id)sharedManager
{
    static VideoRecordingManager *sharedMyManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedMyManager = [[self alloc] init];
    });
    return sharedMyManager;
}

#pragma mark - Camera feed recording using OpenGL

// This is a javascriptAPI-triggered function that starts recording the video feed from the camera background
// The objectKey and IP are used to save the resulting video file at a certain location on a Reality Server
// TODO: some optimization of startRecording and writeFrame should remove the slight lag while recording
- (void)startRecording:(NSString *)objectKey ip:(NSString *)objectIP port:(NSString *)objectPort
{
    CGSize videoOutputSize = CGSizeMake(640, 360); // change this to compress the video to a smaller size. can go up to 1080p.
    float frameRate = 30; // change this to compress the video by recording more/less frames per second
    
    if (isRecording) {
        NSLog(@"Can't record another until first finishes ... already recording");
        return;
    }
    
    NSError *error = nil;

    // generates a random filename and saves to temp file directory before uploading to server
    NSString *videoOutPath = [temporaryDirectory stringByAppendingPathComponent:[[NSString stringWithFormat:@"%u", arc4random() % 10000] stringByAppendingPathExtension:@"mp4"]];
    
    // uses AVAssetWriter to write the camera stream to an mp4 file
    self.assetWriter = [AVAssetWriter assetWriterWithURL:[NSURL fileURLWithPath:videoOutPath] fileType:AVFileTypeMPEG4 error:&error];
    self.assetWriter.shouldOptimizeForNetworkUse = YES;

    NSDictionary *compressionProperties = @{AVVideoProfileLevelKey         : AVVideoProfileLevelH264BaselineAutoLevel,
                                            AVVideoH264EntropyModeKey      : AVVideoH264EntropyModeCABAC,
                                            AVVideoAverageBitRateKey       : @(videoOutputSize.width * videoOutputSize.height * 11.4),
                                            AVVideoMaxKeyFrameIntervalKey  : @60,
                                            AVVideoAllowFrameReorderingKey : @NO};
    
    NSDictionary *videoSettings = @{AVVideoCompressionPropertiesKey : compressionProperties,
                                    AVVideoCodecKey                 : AVVideoCodecTypeH264,
                                    AVVideoWidthKey                 : [NSNumber numberWithFloat:videoOutputSize.width],
                                    AVVideoHeightKey                : [NSNumber numberWithFloat:videoOutputSize.height]};
    
    // creates the asset writer input to handle adding the video frames with the specified compression properties
    self.assetWriterInput = [AVAssetWriterInput assetWriterInputWithMediaType:AVMediaTypeVideo outputSettings:videoSettings];
    
    [self.assetWriterInput setMediaTimeScale:60];
    [self.assetWriter setMovieTimeScale:60];
    [self.assetWriterInput setExpectsMediaDataInRealTime:YES];
    
    NSDictionary *sourcePixelBufferAttributesDictionary = [NSDictionary dictionaryWithObjectsAndKeys:
                                                           [NSNumber numberWithInt:kCVPixelFormatType_32ARGB], kCVPixelBufferPixelFormatTypeKey, nil];
    
    // the pixel buffer input is used to manually write frames into the video using pixel buffers, rather than letting something like ReplayKit add the frames
    self.assetWriterPixelBufferInput = [AVAssetWriterInputPixelBufferAdaptor assetWriterInputPixelBufferAdaptorWithAssetWriterInput:self.assetWriterInput sourcePixelBufferAttributes:sourcePixelBufferAttributesDictionary];
    
    [self.assetWriter addInput:self.assetWriterInput];
    
    recordingStartTime = CACurrentMediaTime(); // frame timestamps are relative to when the video starts
    firstFrameOffset = -1; // the first frame might not get written immediately, so this offsets all the frame times by the difference (-1 means unset)
    
    // actually start the writing session
    [self.assetWriter startWriting];
    [self.assetWriter startSessionAtSourceTime:kCMTimeZero];
    isRecording = true;
    
    // notify the delegate class that recording started, which is responsible for providing pixel buffers each frame via getVideoBackgroundPixels
    [self.videoRecordingDelegate recordingStarted];
    
    // start a loop to add frames to the video at the specified frame rate until stopVideoRecording is called
    updateTimer = [NSTimer scheduledTimerWithTimeInterval:(1.0/frameRate)
                                     target:self
                                   selector:@selector(writeFrame)
                                   userInfo:nil
                                    repeats:YES];
    
    // save the object ID and IP so we can upload to correct server when it finishes
    self.objectID = objectKey;
    self.objectIP = objectIP;
    self.objectPort = objectPort;
}

// this gets called at a fixed interval between startVideoRecording and stopVideoRecording
// it gets a pixel buffer from the videoRecordingDelegate and writes it into the asset writer at the current timestamp
- (void)writeFrame
{
    if (!isRecording) {
        return;
    }
    
    // get the time offset of how long it was between startVideoRecording and the first frame being added
    if (firstFrameOffset < 0) {
        firstFrameOffset = CACurrentMediaTime() - recordingStartTime;
    }
    
    // the first frame will have frameTime = 0, the second will be approximately ~= 0 + (1.0/frameRate), the third ~= 0 + 2 * (1.0/frameRate), ...
    CFTimeInterval frameTime = CACurrentMediaTime() - recordingStartTime - firstFrameOffset;
    
    // need to drop frames if the asset writer isn't ready
    if (!self.assetWriterInput.readyForMoreMediaData) {
        NSLog(@"adaptor not ready %f\n", frameTime);
        return;
    }
    
    // this actually gets an array of pixels in GL_32ARGB format from the AR class responsible for handling the camera
    GLchar* pixels = [self.videoRecordingDelegate getVideoBackgroundPixels];
    
    // this is the size of the raw video feed. this will be compressed to the size specified in AVAssetWriter's outputSettings
    CGSize size = [self.videoRecordingDelegate getCurrentARViewBoundsSize];
    
    // get the pixel buffer pool // TODO: I'm not sure if this is necessary anymore because I changed how I'm writing to the asset writer
    CVPixelBufferPoolRef pixelBufferPool = self.assetWriterPixelBufferInput.pixelBufferPool;
    if (!pixelBufferPool) {
        NSLog(@"Pixel buffer asset writer input did not have a pixel buffer pool available; cannot retrieve frame");
        return;
    }
    
    // create a pixel buffer from the pixel byte array returned by getVideoBackgroundPixels
    CVPixelBufferRef pixelBuffer = NULL;
    GLboolean status = CVPixelBufferCreateWithBytes(kCFAllocatorDefault, size.width, size.height, kCVPixelFormatType_32ARGB, (void*)pixels, size.width*4, NULL, 0, NULL, &pixelBuffer);
    if (status != kCVReturnSuccess) {
        NSLog(@"Could not get pixel buffer from asset writer input; dropping frame...");
        return;
    }
    
    // add the resulting pixel buffer to the asset writer at the current timestamp
    CMTime presentationTime = CMTimeMakeWithSeconds(frameTime, 240);
    [self.assetWriterPixelBufferInput appendPixelBuffer:pixelBuffer withPresentationTime:presentationTime];
}

// This is a javascriptAPI-triggered function that stops the recording that is currently active
// The resulting video file is uploaded to the Reality Server specified by the objectKey and IP provided when startVideoRecording was called
- (void)stopRecording:(NSString *)videoId
{
    if (!isRecording) {
        return;
    }
    
    isRecording = false;
    [self.videoRecordingDelegate recordingStopped];
    
    // stop the writeFrame function from being called on loop
    [updateTimer invalidate];
    
    [self.assetWriterInput markAsFinished];
    
    // when the video finishes writing to disk (at path self.assetWriter.outputURL), upload it to the Reality Server (POST /object/:objectID/video/:videoID)
    [self.assetWriter finishWritingWithCompletionHandler:^{
        NSString* urlEndpoint = [NSString stringWithFormat:@"http://%@:%@/object/%@/video/%@", self.objectIP, self.objectPort, self.objectID, videoId];
        [[FileManager sharedManager] uploadFileFromPath:self.assetWriter.outputURL toURL:urlEndpoint];

        self.assetWriterInput = nil;
        self.assetWriter = nil;
        self.screenRecorder = nil;
    }];
}

#pragma mark - Screen recording using ReplayKit
// This set of start/stop recording is currently NOT used, and there is no way to trigger it from the javascriptAPI right now
// It uses Apple's ReplayKit to record the full screen, *including AR elements and other UI*, instead of just the camera feed
// It works correctly, but interferes with trying to use the device's native screen recording feature at the same time.
// Use startRecording and stopRecording instead, which only record the camera background, not the AR elements.

// Source: https://github.com/anthonya1999/ReplayKit-iOS11-Recorder/blob/master/ReplayKit-iOS11-Recorder/ViewController.m
// TODO: fix startRecordingWithoutAR instead of using this
- (void)startRecordingWithAR:(NSString *)objectKey ip:(NSString *)objectIP port:(NSString *)objectPort
{
    CGSize videoOutputSize = CGSizeMake(640, 360); // change this to compress the video to a smaller size. can go up to 1080p.

    NSError *error = nil;
    
    // generates a random filename and saves to temp file directory before uploading to server
    NSString *videoOutPath = [temporaryDirectory stringByAppendingPathComponent:[[NSString stringWithFormat:@"%u", arc4random() % 10000] stringByAppendingPathExtension:@"mp4"]];
    self.assetWriter = [AVAssetWriter assetWriterWithURL:[NSURL fileURLWithPath:videoOutPath] fileType:AVFileTypeMPEG4 error:&error];
    
    NSDictionary *compressionProperties = @{AVVideoProfileLevelKey         : AVVideoProfileLevelH264BaselineAutoLevel,
                                            AVVideoH264EntropyModeKey      : AVVideoH264EntropyModeCABAC,
                                            AVVideoAverageBitRateKey       : @(videoOutputSize.width * videoOutputSize.height * 11.4),
                                            AVVideoMaxKeyFrameIntervalKey  : @60,
                                            AVVideoAllowFrameReorderingKey : @NO};
    
    NSDictionary *videoSettings = @{AVVideoCompressionPropertiesKey : compressionProperties,
                                    AVVideoCodecKey                 : AVVideoCodecTypeH264,
                                    AVVideoWidthKey                 : [NSNumber numberWithFloat:videoOutputSize.width],
                                    AVVideoHeightKey                : [NSNumber numberWithFloat:videoOutputSize.height]};
    
    self.assetWriterInput = [AVAssetWriterInput assetWriterInputWithMediaType:AVMediaTypeVideo outputSettings:videoSettings];
    
    [self.assetWriter addInput:self.assetWriterInput];
    [self.assetWriterInput setMediaTimeScale:60];
    [self.assetWriter setMovieTimeScale:60];
    [self.assetWriterInput setExpectsMediaDataInRealTime:YES];
    
    self.screenRecorder = [RPScreenRecorder sharedRecorder];
    
    isRecording = true;
    
    [self.screenRecorder startCaptureWithHandler:^(CMSampleBufferRef  _Nonnull sampleBuffer, RPSampleBufferType bufferType, NSError * _Nullable error) {
        if (CMSampleBufferDataIsReady(sampleBuffer)) {
            if (self.assetWriter.status == AVAssetWriterStatusUnknown) {
                [self.assetWriter startWriting];
                [self.assetWriter startSessionAtSourceTime:CMSampleBufferGetPresentationTimeStamp(sampleBuffer)];
            }
            
            if (self.assetWriter.status == AVAssetWriterStatusFailed) {
                NSLog(@"An error occured.");
                return;
            }
            
            if (bufferType == RPSampleBufferTypeVideo) {
                if (self.assetWriterInput.isReadyForMoreMediaData) {
                    [self.assetWriterInput appendSampleBuffer:sampleBuffer];
                }
            }
        }
    } completionHandler:^(NSError * _Nullable error) {
        if (!error) {
            NSLog(@"Recording started successfully.");
            
            // save the object ID and IP and port so we can upload to correct server when it finishes
            self.objectID = objectKey;
            self.objectIP = objectIP;
            self.objectPort = objectPort;
        }
    }];
}

// Stops the recording started with startRecordingWithAR, and uploads the result to the server specified when startRecordingWithAR was called.
// TODO: fix stopRecordingWithoutAR instead of using this
- (void)stopRecordingWithAR:(NSString *)videoId
{
    [self.screenRecorder stopCaptureWithHandler:^(NSError * _Nullable error) {
        if (!error) {
            NSLog(@"Recording stopped successfully. Cleaning up...");
            [self.assetWriterInput markAsFinished];
            [self.assetWriter finishWritingWithCompletionHandler:^{

                // in addition to IP and ID, port can be 8080 or 49369 depending on the server so we store it in another parameter
                NSString* urlEndpoint = [NSString stringWithFormat:@"http://%@:%@/object/%@/video/%@", self.objectIP, self.objectPort, self.objectID, videoId];
                [[FileManager sharedManager] uploadFileFromPath:self.assetWriter.outputURL toURL:urlEndpoint];

                self.assetWriterInput = nil;
                self.assetWriter = nil;
                self.screenRecorder = nil;
            }];
        }
    }];
}

@end

File: Vuforia Spatial Toolbox/VideoRecordingManager.swift
//
//  VideoRecordingManager.swift
//  vst-swift
//
//  Created by Ben Reynolds on 9/23/21.
//

import Foundation
import UIKit
import AVFoundation
import VideoToolbox

protocol VideoRecordingSource {
    var isRecording: Bool { get set }
    
    func getVideoBackgroundPixels() -> [GLchar]
    func getCameraFrameImage() -> UnsafeMutablePointer<VuImageInfo>?
    func getCurrentARViewBoundsSize() -> CGSize

}

extension UIImage {
    public convenience init?(pixelBuffer: CVPixelBuffer) {
        var cgImage: CGImage?
        VTCreateCGImageFromCVPixelBuffer(pixelBuffer, options: nil, imageOut: &cgImage)

        guard let cgImage = cgImage else {
            return nil
        }

        self.init(cgImage: cgImage)
    }
}

class VideoRecordingManager {
    
    // MARK: - Singleton
    static let shared = VideoRecordingManager()
    
    // MARK: - Initialization
    var assetWriter: AVAssetWriter?
    var assetWriterInput: AVAssetWriterInput?
    var pixelBufferInput: AVAssetWriterInputPixelBufferAdaptor?
    
    var isRecording = false
    var recordingStartTime: CFTimeInterval?
    var firstFrameOffset: CFTimeInterval?
//    var updateTimer: Timer?
    
    var displayLink: CADisplayLink?
    
    var objectIP: String?
    var objectID: String?
    var objectPort: String?
    
    var recordingDelegate: VideoRecordingSource?
    
    private init() {
        print("Init FileManager");
    }
    
    func startRecording(objectKey: String, ip: String, port: String) {
        print("startRecording!")
//        let videoOutputSize = CGSize(width: 640, height: 480)
        let videoOutputSize = CGSize(width: 1920, height: 1080)
        let frameRate = 30
        
        if isRecording {
            print("can't record until first finishes ... already recording")
        }
        
        let videoId = "\(Int.random(in: 1 ..< 10000000)).mp4"
        let videoOutPath = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(videoId) //else { return }
        
        print("videoOutPath: \(videoOutPath)")
        
        do {
            assetWriter = try AVAssetWriter(outputURL: videoOutPath, fileType: .mp4)
        } catch {
            print(error)
            return
        }
        
        guard let _writer = assetWriter else { return }
        
        _writer.shouldOptimizeForNetworkUse = true
        _writer.movieTimeScale = 60

        let compressionProperties: [String: Any] = [
            AVVideoProfileLevelKey:             AVVideoProfileLevelH264BaselineAutoLevel,
            AVVideoH264EntropyModeKey:          AVVideoH264EntropyModeCABAC,
            AVVideoAverageBitRateKey:           videoOutputSize.width * videoOutputSize.height * 11.4,
            AVVideoMaxKeyFrameIntervalKey:      60,
            AVVideoAllowFrameReorderingKey:     false
        ]
        
        let videoSettings: [String: Any] = [
            AVVideoCompressionPropertiesKey:    compressionProperties,
            AVVideoCodecKey:                   AVVideoCodecType.h264,
            AVVideoWidthKey:                   videoOutputSize.width,
            AVVideoHeightKey:                  videoOutputSize.height
        ]

        // creates the asset writer input to handle adding the video frames with the specified compression properties

        assetWriterInput = AVAssetWriterInput(mediaType: AVMediaType.video, outputSettings: videoSettings)
        guard let _input = assetWriterInput else { return }

        _input.mediaTimeScale = 60
        _input.expectsMediaDataInRealTime = true
        _input.transform = CGAffineTransform(rotationAngle: CGFloat.pi)
        
        // the pixel buffer input is used to manually write frames into the video using pixel buffers, rather than letting something like ReplayKit add the frames
        
        let sourcePixelBufferAttributes: [String: Any] = [
//            String(kCVPixelBufferPixelFormatTypeKey): Int(kCVPixelFormatType_32ARGB)
            String(kCVPixelBufferPixelFormatTypeKey): Int(kCVPixelFormatType_24RGB)
        ]
        pixelBufferInput = AVAssetWriterInputPixelBufferAdaptor(assetWriterInput: _input, sourcePixelBufferAttributes: sourcePixelBufferAttributes)
        
        _writer.add(_input)

        recordingStartTime = CACurrentMediaTime() // frame timestamps are relative to when the video starts
        firstFrameOffset = -1 // the first frame might not get written immediately, so this offsets all the frame times by the difference (-1 means unset)
        
        _writer.startWriting()
        _writer.startSession(atSourceTime: CMTime.zero)
        
        isRecording = true
        
        // notify the delegate class that recording started, which is responsible for providing pixel buffers each frame via getVideoBackgroundPixels
        // this makes it continually refresh its pixel buffer until recordingStopped() is called
        recordingDelegate?.isRecording = true
        
        // this will trigger at our specified FPS
//        updateTimer = Timer(timeInterval: (1.0 / Double(frameRate)), target: self, selector: #selector(writeFrame), userInfo: nil, repeats: true)
//        RunLoop.current.add(updateTimer, forMode: .common)
        
        // this will trigger each time the screen updates (60 or 120 fps)
        displayLink = CADisplayLink(target: self, selector: #selector(writeFrame))
        displayLink?.add(to: .current, forMode: .default)
        
        // save object id, ip, port for when video finishes writing
        objectIP = ip;
        objectID = objectKey;
        objectPort = port;
    }
    
    @objc func writeFrame() {
        if !isRecording {
            return
        }
        
        if firstFrameOffset == nil {
            firstFrameOffset = CACurrentMediaTime() - (recordingStartTime ?? CACurrentMediaTime())
        }
        
        guard let offset = firstFrameOffset,
              let startTime = recordingStartTime,
//              let _writer = assetWriter,
              let _input = assetWriterInput else { return }
        
        let frameTime = CACurrentMediaTime() - startTime - offset
        
        if !_input.isReadyForMoreMediaData {
            print("_input not ready \(frameTime)")
            return
        }

        guard let cameraImageInfoPointer = recordingDelegate?.getCameraFrameImage() else {
            print("error getting cameraImageInfo")
            cVuforiaCleanupStateMemory()
            return
        }
        
//        print("got cameraImageInfoPointer")
//        print(cameraImageInfoPointer)

        let cameraImageInfo = cameraImageInfoPointer.pointee
//        print(cameraImageInfo)
        
        let pixels: UnsafeMutableRawPointer = UnsafeMutableRawPointer(mutating: cameraImageInfo.buffer)

        print(pixelBufferInput?.pixelBufferPool)
        
        guard (pixelBufferInput?.pixelBufferPool) != nil else {
            cVuforiaCleanupStateMemory()
            return
        }

        var pixelBuffer: CVPixelBuffer? = nil
        
        // Define a function to call when the pixel buffer is freed.
//        let releaseCallback: CVPixelBufferReleaseBytesCallback = { releaseRefCon, baseAddress in
//            print("CVPixelBufferReleaseBytesCallback (1)")
//
//            guard let baseAddress = baseAddress else { return }
//            free(UnsafeMutableRawPointer(mutating: baseAddress))
//            // Perform additional cleanup as needed.
//        }

        print("||--- create buffer")
//        let status = CVPixelBufferCreateWithBytes(kCFAllocatorDefault, Int(cameraImageInfo.width), Int(cameraImageInfo.height), kCVPixelFormatType_24RGB, pixels, Int(cameraImageInfo.stride), { releaseContext, baseAddress in
//            print("CVPixelBufferReleaseBytesCallback (1)")
//        }, nil, nil, &pixelBuffer)
        
        let status = CVPixelBufferCreateWithBytes(kCFAllocatorDefault, Int(cameraImageInfo.width), Int(cameraImageInfo.height), kCVPixelFormatType_24RGB, pixels, Int(cameraImageInfo.stride), nil, nil, nil, &pixelBuffer)

        if status != kCVReturnSuccess {
            print("could not get pixel buffer from asset writer input; dropping frame...")
            cVuforiaCleanupStateMemory()
            return
        }
        
        print("--- buffer created ---||")
        
        let presentationTime = CMTimeMakeWithSeconds(frameTime, preferredTimescale: 240)
        print(presentationTime.seconds)
        
        if let _pixelBuffer = pixelBuffer {
//            if CVPixelBufferLockBaseAddress(_pixelBuffer, CVPixelBufferLockFlags.readOnly) != kCVReturnSuccess {
//                print("error locking base address of pixel buffer")
//                return
//            }
            if let _pixelBufferInput = pixelBufferInput {
                print(">>> append buffer")
                _pixelBufferInput.append(_pixelBuffer, withPresentationTime: presentationTime)
                print("... buffer appended")
                
                cVuforiaCleanupStateMemory()
            }
//            CVPixelBufferUnlockBaseAddress(_pixelBuffer, .readOnly)
        }
        
//        if let _pixelBuffer = pixelBuffer {
//            if CVPixelBufferLockBaseAddress(_pixelBuffer, CVPixelBufferLockFlags.readOnly) != kCVReturnSuccess {
//                print("error locking base address of pixel buffer")
//                return
//            }
//
//            guard let sourceData = CVPixelBufferGetBaseAddress(_pixelBuffer) else {
//                print("could not get pixel buffer base address")
//                CVPixelBufferUnlockBaseAddress(_pixelBuffer, .readOnly)
//                return
//            }
//
//            let sourceBytesPerRow = CVPixelBufferGetBytesPerRow(_pixelBuffer)
//            let sourceWidth = CVPixelBufferGetWidth(_pixelBuffer)
//            let sourceHeight = CVPixelBufferGetHeight(_pixelBuffer)
////            let offset = 0
//
//            guard let scaledData = malloc(sourceHeight * sourceBytesPerRow) else {
//                print("error: out of memory")
//                CVPixelBufferUnlockBaseAddress(_pixelBuffer, .readOnly)
//                return
//            }
//
//            let pixelFormatType = CVPixelBufferGetPixelFormatType(_pixelBuffer)
//            var outputPixelBuffer: CVPixelBuffer? = nil
//
////            let status = CVPixelBufferCreateWithBytes(nil, sourceWidth, sourceHeight, pixelFormatType, scaledData, sourceBytesPerRow, nil, nil, nil, &outputPixelBuffer)
//            let status = CVPixelBufferCreateWithBytes(nil, sourceWidth, sourceHeight, pixelFormatType, sourceData, sourceBytesPerRow,
//              { releaseContext, baseAddress in
//                  print("CVPixelBufferReleaseBytesCallback (2)")
//              }, nil, nil, &outputPixelBuffer)
//
//            CVPixelBufferUnlockBaseAddress(_pixelBuffer, .readOnly)
//
//            if status != kCVReturnSuccess {
//                print("error: could not create new pixel buffer")
//                free(scaledData)
//                return
//            }
//
//            if let _outputBuffer = outputPixelBuffer {
//                if let _pixelBufferInput = pixelBufferInput {
//                    _pixelBufferInput.append(_outputBuffer, withPresentationTime: presentationTime)
//                }
//            }
            
//            let image = CIImage(cvPixelBuffer: pixelBuffer)

//            if let _pixelBufferInput = pixelBufferInput {
////                CVPixelBufferLockBaseAddress(_pixelBuffer, CVPixelBufferLockFlags.readOnly)
//
//                _pixelBufferInput.append(_pixelBuffer, withPresentationTime: presentationTime)
////                print("appended _pixelBuffer to pixelBufferInput")
//
////                CVPixelBufferUnlockBaseAddress(_pixelBuffer, .readOnly)
//            }
//        }
        
        print("writeFrame ended")
    }
    
    func pixelBufferReleaseCallback(releaseRefCon: UnsafeMutableRawPointer?, baseAddress: UnsafeMutableRawPointer?) {
        if baseAddress != nil {
            free(baseAddress)
        }
    }
    
    func stopRecording(videoID: String) {
        if !isRecording {
            return
        }
        
        guard let _writer = assetWriter,
              let _input = assetWriterInput else { return }
        
        isRecording = false
        recordingDelegate?.isRecording = false
        displayLink?.invalidate()
        _input.markAsFinished()
        _writer.finishWriting { [self] in
            if let _ip = objectIP,
               let _id = objectID,
               let _port = objectPort {
                
                if let urlEndpoint = URL(string: "http://\(_ip):\(_port)/object/\(_id)/video/\(videoID)") {
                    FileDownloadManager.shared.uploadVideoFileFromPath(_writer.outputURL, toURL: urlEndpoint)
                } else {
                    print("error creating urlEndpoint for video.. http://\(_ip):\(_port)/object/\(_id)/video/\(videoID)")
                }
            }
            
            assetWriterInput = nil
            assetWriter = nil
            pixelBufferInput = nil
        }
    }
    
    func getScreenshot() -> String? {
        guard let cameraImageInfoPointer = recordingDelegate?.getCameraFrameImage() else {
            print("error getting cameraImageInfo")
            cVuforiaCleanupStateMemory()
            return nil
        }
        
        let cameraImageInfo = cameraImageInfoPointer.pointee
        print(cameraImageInfo)
        
        let pixels: UnsafeMutableRawPointer = UnsafeMutableRawPointer(mutating: cameraImageInfo.buffer)

        var pixelBuffer: CVPixelBuffer? = nil
        
        let status = CVPixelBufferCreateWithBytes(kCFAllocatorDefault, Int(cameraImageInfo.width), Int(cameraImageInfo.height), kCVPixelFormatType_24RGB, pixels, Int(cameraImageInfo.stride), nil, nil, nil, &pixelBuffer)

        if status != kCVReturnSuccess {
            print("could not get pixel buffer from asset writer input; dropping frame...")
            cVuforiaCleanupStateMemory()
            return nil
        }
                
        if let _pixelBuffer = pixelBuffer {
            CVPixelBufferLockBaseAddress(_pixelBuffer, .readOnly)

            let image = CIImage(cvPixelBuffer: _pixelBuffer)
            print(image)
            
            let tempContext = CIContext.init(options: nil)
            if let videoImage = tempContext.createCGImage(image, from: CGRect(x: 0, y: 0, width: CVPixelBufferGetWidth(_pixelBuffer), height: CVPixelBufferGetHeight(_pixelBuffer))) {
                let newUiImage = UIImage(cgImage: videoImage, scale: 0.5, orientation: .down)
                
                let imageData = newUiImage.jpegData(compressionQuality: 0.7)
                let encodedString = imageData?.base64EncodedString(options: .endLineWithCarriageReturn)
                
                CVPixelBufferUnlockBaseAddress(_pixelBuffer, .readOnly)

                cVuforiaCleanupStateMemory()
                
                return encodedString
            }
            
            CVPixelBufferUnlockBaseAddress(_pixelBuffer, .readOnly)
        }

        cVuforiaCleanupStateMemory()
        return nil
    }
    
//    private func imageWith(newSize: CGSize) -> UIImage {
//        let image = UIGraphicsImageRenderer(size: newSize).image { _ in
//            draw(in: CGRect(origin: .zero, size: newSize))
//        }
//
//        return image.withRenderingMode(renderingMode)
//    }
    
    func imageWithImage(_ image: UIImage?, scaledToSize newSize: CGSize) -> UIImage? {
        UIGraphicsBeginImageContextWithOptions(newSize, false, 0.0);
        image?.draw(in: CGRect(x: 0.0, y: 0.0, width: newSize.width, height: newSize.height))
        let newImage: UIImage? = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return newImage
    }
}
